### PR TITLE
mv: gnu test case `to-symlink` fix

### LINKED ISSUE #6577: mv: gnu test case `to-symlink` compatibility
Currently, the GNU test case to-symlink for mv is failing. This test case checks whether mv unlinks the destination symlink before moving a file to a symlink in another partition. If it does not unlink, then the "copy code" used to copy the file to another partition would follow the symlink and replace the file it's pointing to instead of the symlink itself.
**for example**
```shell
echo local > file
mkdir -p  /dev/shm/test/
echo remote > /dev/shm/test/rem_file
ln -rs /dev/shm/test/rem_file /dev/shm/test/rem_symlink
mv file /dev/shm/test/rem_symlink
ls  -l /dev/shm/test/
``` 
should give us 
```
total 8
-rw-r--r-- 1 mhead mhead 6 Jul 17 21:01 rem_file
-rw-r--r-- 1 mhead mhead 6 Jul 17 21:03 rem_symlink
``` 
notice that the `rem_symlink` is now a file and if we cat `rem_symlink` it would have the contents of `file`
but in our case
```
total 4
-rw-r--r-- 1 mhead mhead 6 Jul 17 21:01 rem_file
lrwxrwxrwx 1 mhead mhead 8 Jul 17 20:59 rem_symlink -> rem_file
``` 
`rem_symlink` is still a symlink, what happened is uu_mv followed the symlink  replaced the `rem_file` instead of `rem_symlink`

### PR DESCRIPTION
**Pull request by matrixhead**
tries to fix #6577

**Behaviors changed**
- mv would remove the destination file first, if we are trying to move the file between different partitions and the destination is a symlink

### CODE REVIEW

File: src/uu/mv/src/mv.rs [hunk 1]
```diff
@@ -661,6 +661,9 @@ fn rename_with_fallback(
                 };
             }
         } else {
+            if to.is_symlink() {
+                fs::remove_file(to)?;
```
Comment (BenWiederhake) [inline src/uu/mv/src/mv.rs:665]: If this fails, then the error `Operation not permitted` is very unhelpful. Can you improve upon that? (Also, please add a test that this doesn't regress in the future.)

Example on a system where the current directory is on a different partition than `/tmp`, and the current user is NOT root:

```console
$ echo "src content" > srcfile
$ echo "tmp content" > /tmp/tmpfile
$ sudo ln -s tmpfile /tmp/symlink_to_tmpfile # Create a symlink that cannot be removed
$ mv -v srcfile /tmp/symlink_to_tmpfile 
mv: inter-device move failed: 'srcfile' to '/tmp/symlink_to_tmpfile'; unable to remove target: Operation not permitted
[$? = 1]
$ cargo run -q mv -v srcfile /tmp/symlink_to_tmpfile 
mv: Operation not permitted (os error 1)
[$? = 1]
```

**Review by BenWiederhake — COMMENTED** [review]
Code looks fine, and the test would detect most types of bugs. Great!

The remaining CI failures are unrelated issues: GNU test `timeout` is flaky, #6534, #6570 (both).

Can I ask you to improve the error message though? Right now it's very cryptic. Also, the test could use an additional check, just to make extra sure that we will never accidentally overwrite `other_fs_file.txt` or something.

File: src/uu/mv/src/mv.rs (hunk 1 above)
Comment (matrixhead) [reply src/uu/mv/src/mv.rs:665]: thank you for guidance, I will do this

File: tests/by-util/test_mv.rs [hunk 2]
```diff
@@ -1629,30 +1629,99 @@ fn test_acl() {
...
+    std::os::unix::fs::symlink(&other_fs_file_path, &symlink_path)
+        .expect("Unable to create symlink_to_file");
+
+    // disable write for the target folder so that when mv tries to remove the
+    // the destination symlink inside the target directory it would fail.
+    std::fs::set_permissions(
+        other_fs_tempdir.path(),
+        std::os::unix::fs::PermissionsExt::from_mode(0o555),
+    )
+    .expect("Unable to set permissions for temp directory");
```
Comment (BenWiederhake) [inline tests/by-util/test_mv.rs:1717]: Clever!

**Comment by BenWiederhake** [comment]
Looks good, the only remaining CI failure is unrelated: #6534.

I'm a bit unhappy about the many fully-qualified usages of various traits/structs, like `std::io::Write::write_all` and `std::os::unix::fs::symlink` etc. However, I couldn't find a different test that has to deal with absolute filenames like this, and I can't come up with a better way to do it, so I just accept it and merge it.

If you can come up with a nicer way in the future, please do create a PR :)

**Comment by matrixhead** [comment]
okay i will try

**Pull request merged**